### ISSUE TITLE
slice_embed, select_embed like diag_embed

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#62849]
## 🚀 Feature

We should add `slice_embed` and `select_embed` like `diag_embed`. Furthermore, we should make all three of these operators primitives w.r.t. autograd.

## Motivation

vmap has problems with `slice_backward` and `select_backward` because they decompose into some vmap-incompatible code that involves copy: https://github.com/pytorch/pytorch/blob/e62189ad698303dfd5d577ee437437182ae77918/aten/src/ATen/native/TensorShape.cpp#L1299-L1303. We should turn these into operators and give them backward formulas.

## Pitch

Add the following APIs:
```
Tensor slice_embed(const Tensor& slice, IntArrayRef size, int64_t dim, int64_t start, int64_t end, int64_t step);
Tensor select_embed(const Tensor& selection, IntArrayRef size, int64_t dim, int64_t index);
```
`select_embed` embeds the `selection` tensor in a new zeros Tensor of size `size`

Furthermore, write a backward formula for `slice_embed`, `select_embed`, and `diag_embed`.

## Alternatives

We could just write a backward formula for `slice_backward` and `select_backward` since those are already operators.

**Comment by albanD** [comment]
Do I remember correctly that from a high level point of view, for every operation "foo" that is a view, we want:
- "foo" that is the view op, both as a method and a function
- "foo_copy" that is the same as foo but return brand new memory
- "foo_embed" (do we have other names?) that is an out of place version of `a.foo().copy_()` to be used both as out of place variant for these patterns and for all the backward formulas of the view function that are not allowed to modify the given gradients inplace.

**Comment by albanD** [comment]
By the way, if this is as mechanical as that, this is most likely something we want to do automatically in the codegen! so that the amount of work is not linear with the number of view ops!

**Comment by zou3519** [comment]
We only need `foo_embed` if the view function views a subset of a Tensor's elements. Otherwise, the backward function is the inverse of the view and does not need to be expressed with `copy_`.

cc @bdhirsh -- IIRC foo_copy and "inverse views" are important for the functionalization pass

**Comment by zou3519** [comment]
> "foo_embed" (do we have other names?) that is an out of place version of a.foo().copy_() to be used both as out of place variant for these patterns and for all the backward formulas of the view function that are not allowed to modify the given gradients inplace.

`foo_embed` is actually an out-of-place version of `torch.zeros(shape).foo().copy_(b)`

From the discussions we had a while back, we might also need a `foo_scatter` (I remember @ezyang mentioning these would be useful for functionalization). `foo_scatter` is an out-of-place version of `a.foo().copy_(b)`

NB: Yes, foo_embed can be written as a special case of foo_scatter applied on an all-zeros tensor. But for performance reasons we would want to use foo_embed because foo_scatter applies a copy.

**Comment by ezyang** [comment]
> NB: Yes, foo_embed can be written as a special case of foo_scatter applied on an all-zeros tensor. But for performance reasons we would want to use foo_embed because foo_scatter applies a copy.

If only we had compute tensors... 👀

**Comment by zou3519** [comment]
@albanD would you be fine if we added `slice_embed` and `select_embed` and made `{select_embed, slice_embed, diag_embed}` all primitives w.r.t. autograd and then performed the code generation for the rest later? The code generation is more applicable for the `_copy` and the "_scatter" (inverse?) variants.

Another question is, should these be in Python and should we have documentation for them? I'm leaning towards yes (diag_embed is an op, and these are more generally useful if a user wants to embrace no-mutations)

**Comment by albanD** [comment]
If these are the only two ones you need for now sure.
But if we get more, I think it will be just faster to do the codegen :p
